Environment variables configurable thru yaml file

vkerkhoff opened the pull request:
### Description
Make the RELATIONSHIP/ROUTES/APPLICATION and VARIABLES configurable thru the .magento.app.yaml file to allow users to specify an alternative variable name that is used to get the configuration value.

An example is the use of the ECE tools on a native Platform.sh environment where the variables used are prefixed by PLATFORM instead of MAGENTO_CLOUD.

### Manual testing scenarios
Add for example the following settings in .magento.env.yaml
stage:
  global:
      ENV_RELATIONSHIPS: "PLATFORM_RELATIONSHIPS"
      ENV_ROUTES: "PLATFORM_ROUTES"
      ENV_APPLICATION: "PLATFORM_APPLICATION"
      ENV_VARIABLES: "PLATFORM_VARIABLES"

No automated tests yet, if feature idea is approved I will add them

### Contribution checklist
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds on Travis CI are green)

vkerkhoff committed:
Merge pull request #1 from magento/develop

Update from head

vkerkhoffpon committed:
Environment variables configurable thru yaml file

Make the RELATIONSHIP/ROUTES/APPLICATION and VARIABLES configurable thru the .magento.app.yaml file to allow users to specify an alternative variable name that is used to get the configuration value.

An example is the use of the ECE tools on a native Platform.sh environment where the variables used are prefixed by PLATFORM instead of MAGENTO_CLOUD.

magento-cicd2 commented:
[![CLA assistant check](https://cla.dev.magento.com/pull/badge/signed)](https://cla.dev.magento.com/magento/ece-tools?pullRequest=394) <br/>All committers have signed the CLA.

shiftedreality commented on src/Config/StageConfigInterface.php:
```diff
@@ -36,6 +36,14 @@
 
+    /**
+     * Environment variables.
+     */
```
Could you please move this new varaibles under new config section like:

```
stage: []
system: 
-
-
```

This will make those variables very isolated from stage variables

shiftedreality reviewed (changes requested):
@vkerkhoff thank you for your contribution!
I left 1 comment regarding location of those new varaiables.

shiftedreality commented:
Please also fix failed tests

vkerkhoffpon committed:
Moved variable section to own system element

Added extra section in the .magento.env.yaml that allows storing the new variable options
Created a reader/processor for getting the options from the system section

vkerkhoff commented:
Moved the environment variables options to a new system => variables section in the .magento.env.yaml file. 

Let me know if this setup fits the design wanted so I can also add the tests and check why some tests fail.

vkerkhoffpon committed:
Fixed: Unwanted rename of class

vkerkhoffpon committed:
Fixed: Travis build error

shiftedreality commented:
@vkerkhoff this looks good, thanks!
The only comment I have is about inheritance in `interface VariablesInterface extends SystemConfigInterface`

It looks like `SystemConfigInterface` is independent from stage, so we can avoid this inheritance.

vkerkhoffpon committed:
Removed unneeded VariablesInterface

vkerkhoff commented:
@shiftedreality Thanks! I added the `SystemConfigInterface` so it's possible to add extra sections under the system as like with the stages. I now removed the `VariablesInterface` and only use the SystemConfigInterface

vkerkhoffpon committed:
Fixed: Build error on testGet

shiftedreality commented on src/Config/SystemConfigInterface.php:
```diff
@@ -0,0 +1,44 @@
+    /**
+     * Option for enabling merging given configuration with default configuration
+     */
+    const OPTION_MERGE = '_merge';
```
This seems not necessary in system config

vkerkhoffpon committed:
Removed un-used constant

vkerkhoffpon committed:
Added Mock of depended class

vkerkhoffpon committed:
Added docblock

billygilbert commented on src/Config/Environment.php:
```diff
@@ -68,6 +85,17 @@ public function get(string $key, $default = null)
     }
 
+    /**
+     * Get environment variable and get the name from .magento.env.yaml configuration file
```
Add period at end of comment please.

billygilbert commented on src/Config/Environment.php:
```diff
@@ -31,6 +38,16 @@ class Environment
 
+    /**
+     * Environment constructor.
+     * @param Variables $systemConfig
```
Add new line between 42/43

billygilbert commented on src/Config/Environment.php:
```diff
@@ -31,6 +38,16 @@ class Environment
+     */
+    public function __construct(
+        Variables $systemConfig
+    ) {
```
No need for new lines here.

billygilbert commented on src/Config/SystemConfigInterface.php:
```diff
@@ -0,0 +1,39 @@
+    const SECTION_SYSTEM = 'system';
+
+    /**
+     * System sections
```
Add period at end of comment please.

billygilbert reviewed (changes requested):
Please also update the `dist/.magento.env.yaml` file: https://github.com/magento/ece-tools/blob/develop/dist/.magento.env.yaml

Also for completeness, please also move MAGENTO_CLOUD_MODE and MAGENTO_CLOUD_ENVIRONMENT

Thanks for the contribution!!!

vkerkhoffpon committed:
Added usage documentation
Fixed comments from code review (doc blocks and code styling)
Added MAGENTO_CLOUD_MODE and MAGENTO_CLOUD_ENVIRONMENT variables to be configurable
Updated Unit Tests to cover changes

vkerkhoff commented:
@billygilbert I updated the code with your comments and added some documentation in the dist/.magento.env.yaml file.

@shiftedreality I added/fixed the unit tests, the only message I don't see how to resolve is the code coverage, if this needs some changes please give some input on what to add.

vkerkhoffpon committed:
Fixed double newline

shiftedreality commented:
@vkerkhoff code coverage is increasing by unit tests. Looks like it's already green

vkerkhoff commented:
@shiftedreality Thanks, I see it indeed. Funny that if I run the test locally I get 94.69% code coverage.

shiftedreality commented:
@vkerkhoff I observed the same and do not have a clear explanation of this. Looks like it somehow depends on OS(I use Mac).
Thanks for fixing the tests, I will proceed with QA.

billygilbert commented on dist/.magento.env.yaml:
```diff
@@ -447,3 +447,57 @@
+#          system:                                                                                                    #
+#            variables:                                                                                               #
+#              ENV_ENVIRONMENT: "MAGENTO_CLOUD_ENVIRONMENT"                                                           #
+#######################################################################################################################
```
👍

bbatsche commented on src/Config/Stage/Deploy.php:
```diff
@@ -34,7 +35,7 @@ class Deploy implements DeployInterface
     private $mergedConfig;
 
     /**
-     * @var Environment
```
I believe this type should still be `Environment`

bbatsche commented on src/Config/Stage/Deploy.php:
```diff
@@ -44,7 +45,7 @@ class Deploy implements DeployInterface
     private $schema;
 
     /**
-     * @param Environment $environment
```
Same with this type.

bbatsche commented on src/Config/Stage/Deploy.php:
```diff
@@ -127,8 +128,9 @@ private function mergeConfig(): array
+        $envVar = $this->environment->getEnvVarName(SystemConfigInterface::VAR_ENV_MODE);
 
-        if ($this->environment->getEnv('MAGENTO_CLOUD_MODE')  === Environment::CLOUD_MODE_ENTERPRISE) {
+        if ($this->environment->getEnv($envVar)  === Environment::CLOUD_MODE_ENTERPRISE) {
```
As long as you're here, remove the extra space before `===`

vkerkhoffpon committed:
Updated doc block comments

vkerkhoff commented:
@bbatsche Thanks, I updated the requested changes

shiftedreality commented:
QA approved

shiftedreality commented:
@vkerkhoff thank you for your contribution!
